### PR TITLE
(CDAP-7913) Use TMS for notification.

### PR DESCRIPTION
**Pull request by chtyim**

### CODE REVIEW

**Comment by chtyim** [comment]
Test on cluster and ran the StreamSchedulerTest in ITN

File: cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/MessagingNotificationService.java [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+ * Implementation of the {@link NotificationService} that uses TMS as the transport.
+ */
+public class MessagingNotificationService extends AbstractNotificationService {
+  private static final Logger LOG = LoggerFactory.getLogger(MessagingNotificationService.class);
+
+  private final MessagingService messagingService;
+  private final TopicId notificationTopic;
+  private final AtomicBoolean needFetch;
+
+  // Executor to publish notifications to Kafka
```
Comment (gokulavasan) [inline cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/MessagingNotificationService.java]: notifications to TMS

File: cdap-notifications/src/test/java/co/cask/cdap/notifications/kafka/MessagingNotificationTest.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import co.cask.cdap.notifications.guice.NotificationServiceRuntimeModule;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import com.google.inject.Guice;
+import com.google.inject.Injector;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+
+/**
+ * Tests notifications using Kafka transport system.
```
Comment (gokulavasan) [inline cdap-notifications/src/test/java/co/cask/cdap/notifications/kafka/MessagingNotificationTest.java]: Tests notifications

File: cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/NotificationMessage.java [hunk 3]
```diff
@@ -16,34 +16,35 @@
...
     return notificationJson;
   }
 
   @Override
   public String toString() {
-    return Objects.toStringHelper(this)
-      .add("messageKey", messageKey)
-      .add("notificationJson", notificationJson)
-      .toString();
+    return "KafkaMessage{" +
```
Comment (gokulavasan) [inline cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/NotificationMessage.java]: Why is it still called KafkaMessage?

File: cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/MessagingNotificationService.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+          }
+        } catch (Exception e) {
+          LOG.error("Failed to get notification", e);
+        }
+
+        // Back-off if it was empty fetch.
+        if (emptyFetchCount > 0) {
+          // Schedule the next fetch. Exponential strategy doesn't use the time component,
+          // so doesn't matter what we passed in
+          subscribeExecutor.schedule(this,
```
Comment (gokulavasan) [inline cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/MessagingNotificationService.java:170]: Let's say the first time we try to fetch messages, nothing is available. So we schedule our next execution, but meanwhile, a few notifications are published. How will we fetch those notifications, if we use the System#currentTimeMillis again during the next execution?

File: cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/NotificationMessage.java [hunk 5]
```diff
@@ -16,34 +16,35 @@
 
 package co.cask.cdap.notifications.service.kafka;
 
-import com.google.common.base.Objects;
+import co.cask.cdap.proto.id.NotificationFeedId;
 import com.google.gson.JsonElement;
 
 /**
  * Message sent to Kafka that contains a serialized notification.
```
Comment (gokulavasan) [inline cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/NotificationMessage.java]: Kafka -> TMS

**Comment by gokulavasan** [comment]
Finished review. One question, rest are minor comments.

File: cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/NotificationMessage.java (hunk 3 above)
Comment (chtyim) [reply cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/NotificationMessage.java]: haha... apparently IntelliJ rename class refactoring won't touch the toString :)

File: cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/MessagingNotificationService.java (hunk 4 above)
Comment (chtyim) [reply cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/MessagingNotificationService.java:170]: The `System.currentTimeMillis` is a field, hence it only get set one when this runnable is created. We don't create new runnable instance when schedule again.

Comment (gokulavasan) [reply cdap-notifications/src/main/java/co/cask/cdap/notifications/service/kafka/MessagingNotificationService.java:170]: Ah didn't notice that. Thanks!

**Comment by gokulavasan** [comment]
LGTM 👍

**Comment by chtyim** [comment]
build passed http://builds.cask.co/browse/CDAP-RUT404

**Pull request merged**